A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,184 @@
+field will be ignored as it is only used in the "*all weights not equal*" (weighted least request)
+case which will not be supported as part of this proposal.
+
+#### Changes to the `xds_cluster_resolver` Policy
+
+As a result of
+[gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md),
+the `xds_cluster_resolver` policy implementation will need additional logic to understand the setup for `LEAST_REQUEST`
+in addition to `ROUND_ROBIN` and `RING_HASH`.
+For `LEAST_REQUEST`, the policy name will be `"LEAST_REQUEST"`, and the config will be the one for the `least_request_experimental` LB Policy described below.
+When the `xds_cluster_resolver` policy sees `"LEAST_REQUEST"`, it will also assume that the
+locality-picking policy is `weighted_target` and the endpoint-picking policy is `least_request_experimental`.
+
+#### `least_request_experimental` LB Policy
+
+We will implement a `least_request_experimental` LB policy that uses the same
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+##### LB Policy Config
+
+The `least_request_experimental` policy will have the following config:
+
+```proto
+message LeastRequestLoadBalancingConfig {
+  uint32 choice_count = 1;  // Optional, defaults to 2.
+}
+```
+
+The `choice_count` determines the number of randomly sampled subchannels which should be compared
+for each pick by the picker.
+
+##### Subchannel state handling
+
+The subchannel state handling implemented in the `round_robin` policy will be re-used for `least_request_experimental`.
+Implementations may break out this subchannel state handling and have `round_robin` and `least_request_experimental`
+inherit the logic instead. This would be similar to the "base balancer" already implemented in grpc-go
+[here](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go).
+
+TODO: Possibly flesh this out with more explicit logic pending proposal PR review.
+
+##### Aggregated Connectivity State
+
+The `least_request_experimental` policy will use the same heuristic for determining the aggregated
+connectivity state as `ring_hash_experimental` defined in
+[gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md#aggregated-connectivity-state).
+As explained there, this heuristic ensures that the priority policy will fail over to the
+next priority quickly when there's an outage.
+
+##### Picker Behavior
+
+The picker should be given a list of all subchannels with the `READY` state.
+With this list, the picker should pick a subchannel based on the following pseudo-code:
+
+```
+candidate = null;
+for (int i = 0; i < choiceCount; ++i) {
+    sampled = subchannels[random_integer(0, subchannels.length)]
```

Comment:
super-nit: seems this pseudo-code language uses semicolons, but the semicolon is missing here.